lib/puppet/functions/vault_lookup/lookup.rb

```diff
@@ -2,48 +2,77 @@
   dispatch :lookup do
     param 'String', :path
     optional_param 'String', :vault_url
+    optional_param 'String', :vault_namespace
+    optional_param 'String', :vault_cert_path_segment
+    optional_param 'String', :vault_cert_role
   end
 
-  def lookup(path, vault_url = nil)
+  def lookup(path,
+             vault_url = nil,
+             vault_namespace = nil,
+             vault_cert_path_segment = 'v1/auth/cert/',
+             vault_cert_role = nil)
     if vault_url.nil?
       Puppet.debug 'No Vault address was set on function, defaulting to value from VAULT_ADDR env value'
       vault_url = ENV['VAULT_ADDR']
       raise Puppet::Error, 'No vault_url given and VAULT_ADDR env variable not set' if vault_url.nil?
     end
 
-    uri = URI(vault_url)
-    # URI is used here to just parse the vault_url into a host string
+    vault_base_uri = URI(vault_url)
+    # URI is used here to parse the vault_url into a host string
     # and port; it's possible to generate a URI::Generic when a scheme
     # is not defined, so double check here to make sure at least
     # host is defined.
-    raise Puppet::Error, "Unable to parse a hostname from #{vault_url}" unless uri.hostname
+    raise Puppet::Error, "Unable to parse a hostname from #{vault_url}" unless vault_base_uri.hostname
 
-    use_ssl = uri.scheme == 'https'
-    connection = Puppet::Network::HttpPool.http_instance(uri.host, uri.port, use_ssl)
+    client = Puppet.runtime[:http]
+    token = get_cert_auth_token(client,
+                                vault_base_uri,
+                                vault_namespace,
+                                vault_cert_path_segment,
+                                vault_cert_role)
 
-    token = get_auth_token(connection)
+    secret_uri = vault_base_uri + "/v1/#{path.delete_prefix('/')}"
+    data = get_secret(client, secret_uri, token)
+    Puppet::Pops::Types::PSensitiveType::Sensitive.new(data)
+  end
 
-    secret_response = connection.get("/v1/#{path}", 'X-Vault-Token' => token)
-    unless secret_response.is_a?(Net::HTTPOK)
-      message = "Received #{secret_response.code} response code from vault at #{uri.host} for secret lookup"
-      raise Puppet::Error, append_api_errors(message, secret_response)
+  private
+
+  def auth_login_body(vault_cert_role)
+    if vault_cert_role
+      { name: vault_cert_role }.to_json
+    else
+      ''
     end
+  end
 
+  def get_secret(client, uri, token)
+    secret_response = client.get(uri, headers: { 'X-Vault-Token' => token })
```

Comment:
Since vault namespacing is an enterprise feature and we don't have a good way to stand up a Vault Enterprise instance to ensure this works I am going to remove this for now.